server.js

```diff
@@ -57,8 +131,67 @@ app.post("/user/breached", function(req, res) {
   res.json({ info: "breach alert sent", emails: response });
 });
 
+// This object exists instead of inlining the env vars to make it easy
+// to abstract fetching API endpoints from the OAuth server (instead
+// of specifying them in the environment) in the future.
+const FxAOAuthUtils = {
+  get authorizationUri() { return process.env.OAUTH_AUTHORIZATION_URI },
+  get tokenUri() { return process.env.OAUTH_TOKEN_URI },
+  get profileUri() { return process.env.OAUTH_PROFILE_URI },
+};
+
+var FxAOAuth = new ClientOAuth2({
+  clientId: process.env.OAUTH_CLIENT_ID,
+  clientSecret: process.env.OAUTH_CLIENT_SECRET,
+  accessTokenUri: FxAOAuthUtils.tokenUri,
+  authorizationUri: FxAOAuthUtils.authorizationUri,
+  redirectUri: localServerURL + "/oauth/redirect",
+  scopes: ["profile:email"],
+});
+
+app.get("/oauth/init", function(req, res) {
+  // Set a random state string in a cookie so that we can verify
+  // the user when they're redirected back to us after auth.
+  let state = getStateString();
+  let uri = FxAOAuth.code.getUri({state});
+  req.session.state = state;
+  res.redirect(uri);
+});
+
+app.get('/oauth/redirect', function (req, res) {
+  if (!req.session.state) {
+    res.send("Who are you?");
+    return;
+  }
+  FxAOAuth.code.getToken(req.originalUrl, { state: req.session.state })
+    .catch(function (err) {
+      res.send(err);
+    })
```

Comment:
I think I may need somebody explain this bit to me. IIUC, the `.catch()` statement in the middle of this promise chain will swallow any error that may occur during `FxAOAuth.code.getToken(...)`, but then the `.catch()` will send an error to the user and then continue onto the `.then(function(user) {...})` code path (where `user` will probably be `null|undefined`).

Should the `.catch()` be at the end (after the `.then()`)? Do we need to add more error handling in the case of `popsicle.get(...).then(data => {})` generates an error?

Reply:
Yeah, this is wonky. The catch you're bothered by (and rightly so) was an artifact of me trying to debug why getToken wasn't working (see https://github.com/mozilla/fxa-oauth-server/issues/513). I'll move the catch to after the then() block, but I'll also file an issue about adding real error logging.